### PR TITLE
Implements #607 - Run apply-all only on specified targets

### PR DESCRIPTION
**Pull request by mattiarossi**
This implements #607 by adding a -terragrunt-include-dir CLI flag.

It supports unix-style globs and will include all modules within or under the specified directories during execution. It will ignore any other module (except for dependencies)

I have reused code from the --terragrunt-exclude-dir patch, shouldExcludeModuleBecauseOfPath  would need renaming in order to reflect the dual usage
Also, I reused the TEST_FIXTURE_LOCAL_WITH_EXCLUDE_DIR in the integration test, let me know if you need them duplicated in  a dedicated folder instead

Regards,
Mattia

### CODE REVIEW

File: configstack/module.go [hunk 1]
```diff
@@ -130,6 +135,67 @@ func flagExcludedDirs(modules []*TerraformModule, terragruntOptions *options.Ter
...
+	}
+
+	canonicalWorkingDir, err := util.CanonicalPath("", terragruntOptions.WorkingDir)
+	if err != nil {
+		return nil, err
+	}
+
+	includeGlobMatches := []string{}
+
+	// If possible, expand the glob to get all excluded filepaths
```
Comment (brikis98) [inline configstack/module.go]: Included filepaths?

File: configstack/module.go [hunk 2]
```diff
@@ -130,6 +135,67 @@ func flagExcludedDirs(modules []*TerraformModule, terragruntOptions *options.Ter
...
+	for _, dir := range terragruntOptions.IncludeDirs {
+
+		absoluteDir := ""
+
+		// Ensure excludedDirs are absolute
+		if filepath.IsAbs(dir) {
+			absoluteDir = dir
+		} else {
+			absoluteDir = filepath.Join(canonicalWorkingDir, dir)
+		}
```
Comment (brikis98) [inline configstack/module.go]: Could simplify to:

```go
absoluteDir := dir
if !filepath.IsAbs(absoluteDir) {
  absoluteDir = filepath.Join(canonicalWorkingDir, absoluteDir)
}
```

File: configstack/module.go [hunk 3]
```diff
@@ -130,6 +135,67 @@ func flagExcludedDirs(modules []*TerraformModule, terragruntOptions *options.Ter
...
+	for _, module := range includeGlobMatches {
+		canonicalPath, err := util.CanonicalPath(module, terragruntOptions.WorkingDir)
+		if err != nil {
+			return nil, err
+		}
+		canonicalIncludeDirs = append(canonicalIncludeDirs, canonicalPath)
+	}
+
+	for _, module := range modules {
+		if shouldExcludeModuleBecauseOfPath(module, canonicalIncludeDirs) {
```
Comment (brikis98) [inline configstack/module.go]: Yea, rename this method, as otherwise, this is quite confusing to read

File: configstack/module.go [hunk 4]
```diff
@@ -130,6 +135,67 @@ func flagExcludedDirs(modules []*TerraformModule, terragruntOptions *options.Ter
...
+			// Mark module itself as excluded
+			module.FlagExcluded = false
```
Comment (brikis98) [inline configstack/module.go]: Mark module as included?

File: options/options.go [hunk 5]
```diff
@@ -135,6 +138,7 @@ func NewTerragruntOptions(terragruntConfigPath string) (*TerragruntOptions, erro
 		Sleep:                  DEFAULT_SLEEP,
 		RetryableErrors:        util.CloneStringList(RETRYABLE_ERRORS),
 		ExcludeDirs:            []string{},
+		IncludeDirs:            []string{},
```
Comment (brikis98) [inline options/options.go]: Please update the `Clone` method further down in this file too

File: test/integration_test.go [hunk 6]
```diff
@@ -1236,6 +1236,70 @@ func TestExcludeDirs(t *testing.T) {
...
+func TestIncludeDirs(t *testing.T) {
+	t.Parallel()
+
+	// Populate module paths.
+	moduleNames := []string{
+		"integration-env/aws/module-aws-a",
+		"integration-env/gce/module-gce-b",
+		"integration-env/gce/module-gce-c",
+		"production-env/aws/module-aws-d",
+		"production-env/gce/module-gce-e",
```
Comment (brikis98) [inline test/integration_test.go]: If these folders are re-used from other test cases, and this test runs in parallel in those same folders, then I think you're going to have conflicts. You'll want to make a copy of these folders, either now ahead of time, or at run-time using the copy methods you'll find used elsewhere in `integration_test.go`.

File: configstack/module.go (hunk 1 above)
Comment (mattiarossi) [reply configstack/module.go]: Done

File: configstack/module.go (hunk 2 above)
Comment (mattiarossi) [reply configstack/module.go]: Done

File: configstack/module.go (hunk 3 above)
Comment (mattiarossi) [reply configstack/module.go]: Renamed shouldExcludeModuleBecauseOfPath to findModuleinPath

File: configstack/module.go (hunk 4 above)
Comment (mattiarossi) [reply configstack/module.go]: done, and sorry for overlooking all the comments :)

File: options/options.go (hunk 5 above)
Comment (mattiarossi) [reply options/options.go]: Done

File: test/integration_test.go (hunk 6 above)
Comment (mattiarossi) [reply test/integration_test.go]: I went for the copy, we may need to cerate different test cases to better check the include logic,  so I thought it was better to duplicate test/fixture-download/local-with-exclude-dir to test/fixture-download/local-with-include-dir

**Comment by mattiarossi** [comment]
Pushed a commit with all the fixes, an update to the README and all requested changes

**Comment by brikis98** [comment]
https://github.com/gruntwork-io/terragrunt/releases/tag/v0.17.4

**Comment by brikis98** [comment]
Hm, tests failed on release. Relevant snippets:

```
--- FAIL: TestIncludeDirs (6.59s)
...
	integration_test.go:1310: [TestIncludeDirs] Error copying source module: readdirent: no such file or directory
	integration_test.go:1310: [TestIncludeDirs] [terragrunt] [/go/src/github.com/gruntwork-io/terragrunt/test/fixture-download/local-with-include-dir/integration-env/gce/module-gce-c] 2018/12/21 17:00:48 Module /go/src/github.com/gruntwork-io/terragrunt/test/fixture-download/local-with-include-dir/integration-env/gce/module-gce-c has finished with an error: Hit multiple errors:
	integration_test.go:1310: [TestIncludeDirs] exit status 1
```

**Comment by mattiarossi** [comment]
Weird, it runs locally, but haven't run all tests, doing it now

**Comment by mattiarossi** [comment]
I have tried to run go test -v -parallel 128 in ./test but depending on the OS I run it it fails at different points, and cannot get to TestIncludeDirs:

`
[terragrunt] [/tmp/terragrunt-test535874485/fixture-stack/mgmt/vpc] 2018/12/22 02:30:56 Module /tmp/terragrunt-test535874485/fixture-stack/mgmt/vpc has finished successfully!
[terragrunt] [/tmp/terragrunt-test535874485/fixture-stack/mgmt/kms-master-key] 2018/12/22 02:30:56 Module /tmp/terragrunt-test535874485/fixture-stack/mgmt/kms-master-key has finished successfully!
--- PASS: TestTerragruntStackCommands (161.90s)
    integration_test.go:1363: Copying fixture-stack/ to /tmp/terragrunt-test535874485
    integration_test.go:1556: Deleting test s3 bucket terragrunt-test-bucket-wehrnv
FAIL
exit status 1
FAIL	github.com/gruntwork-io/terragrunt/test	162.795s
`

**Comment by mattiarossi** [comment]
.. and scratch that, testincludedirs passes:
`
--- PASS: TestTerragruntBeforeAndAfterMergeHook (8.88s)
    integration_test.go:255: bucketName: terragrunt-test-bucket-5r2iik
    integration_test.go:1556: Deleting test s3 bucket terragrunt-test-bucket-5r2iik
--- PASS: TestTerragruntWorksWithIncludes (9.08s)
    integration_test.go:1556: Deleting test s3 bucket terragrunt-test-bucket-fhfwej
--- PASS: TestIncludeDirs (9.34s)
`

The fail happens later, just after destroying the temp s3 bucket

**Comment by brikis98** [comment]
Just re-ran the tests and this time they passed... Perhaps there was some intermittent issue? We'll keep our eyes on the tests...

**Comment by stonefury** [comment]
Am I using this wrong?  I want to build any folder underneath me that matches this pattern *mcs*.

te plan-all --terragrunt-include-dir=*mcs*

te plan-all --terragrunt-include-dir="*mcs*"

thanks
